File: rust/src/bin/p0080.rs
// Project Euler: Problem 80

euler::run_solver!(80);

fn solve() -> String {
    compute(100, 100).to_string()
}

fn compute(limit: i64, n_digits: u32) -> i64 {
    use euler::math;
    use num_bigint::BigUint;

    debug_assert!(limit > 1);

    let power_10 = BigUint::from(10_u32).pow((n_digits - 1) * 2);
    (2..=limit)
        .filter(|&x| !math::is_square(x))
        .map(|x| (power_10.clone() * x as u32).sqrt())
        .map(|x| {
            x.to_radix_le(10)
                .into_iter()
                .rev()
                .take(n_digits as usize)
                .map(i64::from)
                .sum::<i64>()
        })
        .sum()
}

#[cfg(test)]
mod tests {
    use super::compute;

    #[test]
    fn p0080_2() {
        assert_eq!(compute(2, 100), 475);
    }

    #[test]
    fn p0080_100() {
        assert_eq!(compute(100, 100), 40886);
    }
}

File: ocaml/src/p0080.ml
(* Project Euler: Problem 80 *)

(*
 * This program needs the Zarith module.
 *
 * when n < 100,
 *
 *   i <= 10^99 * sqrt(n) < i + 1
 *  -->
 *   i^2 <= 10^198 * n < (i + 1)^2
 *
 * 'i' is the 100-digit number we want.
 *)

open Core

let compute upper n_digit =
  let const_pow_of_10 = Z.pow (Z.of_int 10) ((n_digit - 1) * 2) in
  List.(
    range 1 upper ~stop:`inclusive
    |> map ~f:(fun n ->
      if Euler.Math.is_square n
      then 0
      else
        Z.(sqrt (const_pow_of_10 * ~$n))
        |> Euler.Util.z_digits
        |> rev
        |> Fun.flip take n_digit
        |> reduce_exn ~f:( + ))
    |> reduce_exn ~f:( + ))
;;

let solve () = compute 100 100 |> Int.to_string

(* Test *)

let%test_unit "100" = [%test_eq: int] (compute 100 100) 40886
